File: ports/go/src/core/kernel/bcskernel.go
package kernel

import (
	"sync"
	"fmt"
)

// Module interfaces for lifecycle registration
type BcsModule interface {
	Initialize() error
	Start() error
	Shutdown() error
	Name() string
}

type BcsKernel struct {
	*BcsObject
	modules      []BcsModule
	eventLoop    *BcsEventLoop
	mu           sync.Mutex
	isStarted    bool
	isShuttingDown bool
}

var instance *BcsKernel
var once sync.Once

func GetKernel() *BcsKernel {
	once.Do(func() {
		instance = &BcsKernel{
			BcsObject: NewBcsObject(nil),
			modules:   make([]BcsModule, 0),
		}
		// In Go, we explicitly set the global event loop for the main thread here
		instance.eventLoop = NewBcsEventLoop(instance.BcsObject)
	})
	return instance
}

func (k *BcsKernel) RegisterModule(mod BcsModule) {
	k.mu.Lock()
	defer k.mu.Unlock()
	if !k.isStarted {
		k.modules = append(k.modules, mod)
	}
}

func (k *BcsKernel) Startup() error {
	k.mu.Lock()
	if k.isStarted {
		k.mu.Unlock()
		return nil
	}
	k.isStarted = true
	k.mu.Unlock()

	// Phase 1: Initialize all registered modules
	for _, mod := range k.modules {
		fmt.Printf("BcsKernel: Initializing module %s\n", mod.Name())
		if err := mod.Initialize(); err != nil {
			return fmt.Errorf("module %s failed to initialize: %v", mod.Name(), err)
		}
	}

	// Phase 2: Start all initialized modules
	for _, mod := range k.modules {
		fmt.Printf("BcsKernel: Starting module %s\n", mod.Name())
		if err := mod.Start(); err != nil {
			return fmt.Errorf("module %s failed to start: %v", mod.Name(), err)
		}
	}

	return nil
}

func (k *BcsKernel) Exec() int {
	if err := k.Startup(); err != nil {
		fmt.Printf("BcsKernel Startup Error: %v\n", err)
		return -1
	}

	fmt.Println("BcsKernel: Entering main event loop...")
	code := k.eventLoop.Exec()

	k.Shutdown()
	return code
}

func (k *BcsKernel) Shutdown() {
	k.mu.Lock()
	if k.isShuttingDown {
		k.mu.Unlock()
		return
	}
	k.isShuttingDown = true
	k.mu.Unlock()

	fmt.Println("BcsKernel: Shutting down...")

	// Reverse order shutdown
	for i := len(k.modules) - 1; i >= 0; i-- {
		mod := k.modules[i]
		fmt.Printf("BcsKernel: Shutting down module %s\n", mod.Name())
		mod.Shutdown()
	}

	// Finally destroy object tree
	k.Destroy()
}

File: ports/rust/src/core/kernel/bcs_kernel.rs
use std::sync::{Arc, Mutex, OnceLock};
use super::bcs_object::BcsObject;
use super::bcs_event_loop::BcsEventLoop;
use super::bcs_event_dispatcher::BcsEventDispatcher;

pub trait BcsModule: Send + Sync {
    fn initialize(&self) -> Result<(), String>;
    fn start(&self) -> Result<(), String>;
    fn shutdown(&self) -> Result<(), String>;
    fn name(&self) -> String;
}

pub struct BcsKernel {
    pub base: Arc<BcsObject>,
    pub event_loop: Arc<BcsEventLoop>,
    modules: Mutex<Vec<Arc<dyn BcsModule>>>,
    is_started: Mutex<bool>,
    is_shutting_down: Mutex<bool>,
}

static INSTANCE: OnceLock<Arc<BcsKernel>> = OnceLock::new();

impl BcsKernel {
    pub fn instance() -> Arc<BcsKernel> {
        INSTANCE.get_or_init(|| {
            let base = BcsObject::new();
            let dispatcher = Arc::new(BcsEventDispatcher::new());
            let event_loop = Arc::new(BcsEventLoop::new(dispatcher, Some(base.clone())));

            Arc::new(Self {
                base,
                event_loop,
                modules: Mutex::new(Vec::new()),
                is_started: Mutex::new(false),
                is_shutting_down: Mutex::new(false),
            })
        }).clone()
    }

    pub fn register_module(&self, module: Arc<dyn BcsModule>) {
        if !*self.is_started.lock().unwrap() {
            self.modules.lock().unwrap().push(module);
        }
    }

    pub fn startup(&self) -> Result<(), String> {
        let mut started = self.is_started.lock().unwrap();
        if *started {
            return Ok(());
        }
        *started = true;

        let modules = self.modules.lock().unwrap().clone();

        // Phase 1: Initialize
        for module in &modules {
            println!("BcsKernel: Initializing module {}", module.name());
            module.initialize()?;
        }

        // Phase 2: Start
        for module in &modules {
            println!("BcsKernel: Starting module {}", module.name());
            module.start()?;
        }

        Ok(())
    }

    pub fn exec(&self) -> i32 {
        if let Err(e) = self.startup() {
            println!("BcsKernel Startup Error: {}", e);
            return -1;
        }

        println!("BcsKernel: Entering main event loop...");
        let code = self.event_loop.exec();

        self.shutdown();
        code
    }

    pub fn shutdown(&self) {
        let mut shutting_down = self.is_shutting_down.lock().unwrap();
        if *shutting_down {
            return;
        }
        *shutting_down = true;

        println!("BcsKernel: Shutting down...");

        let modules = self.modules.lock().unwrap().clone();

        // Reverse order shutdown
        for module in modules.iter().rev() {
            println!("BcsKernel: Shutting down module {}", module.name());
            let _ = module.shutdown();
        }

        self.base.destroy();
    }
}
